docs/design/mono/mobile-runtimeconfig-json.md

```diff
@@ -0,0 +1,143 @@
+
+#### Task Contract:
+
+The task will take 3 input arguments:
+1. The path to the `runtimeconfig.json` file.
+2. The name of the destination file to be written in the encoded format.
+3. An item list (`ITaskItem[]`) that is the name of the properties that the embedder will set on `monovm_initialize`.
+
+The task should:
+1. Parse the given input file and create a dictionary from the configProperties key.
+2. Compare the keys from the input file with the names given in the item list. If there are any duplicates, return an MSBuild Error.
+3. Generate the output file.
+
+#### Example of the usage of the task:
+
+ ```
+ <UsingTask TaskName="RuntimeConfigParserTask"
+            AssemblyFile="$(RuntimeConfigParserTasksAssemblyPath)" />
+
+<Target Name="BundleTestAndroidApp">
+  <RuntimeConfigParserTask
+      RuntimeConfigFile="$(Path_to_runtimeconfig.json_file)"
+      OutputFile="$(Path_to_generated_binary_file)"
+      RuntimeConfigReservedProperties="@(runtime_properties_reserved_by_host)">
+  </RuntimeConfigParserTask>
+</Target>
+ ```
+
+### The encoded runtimeconfig format
+
+The format is:
+1. There will be an (1- to 4-byte) ECMA-335 II.23.2 compressed unsigned integer count = N, indicating the number of key-value pairs.
+2. It will have 2xN  ECMA-335 SerString UTF8 strings (that is, each string is preceded by its length stored in the compressed unsigned integer format): Each key followed by its value.
+
+Sample input
+```
+{
+    "runtimeOptions": {
+        "configProperties": {
+            "key1": "value1",
+            "key2": "value2"
+        }
+    }
+}
+```
+
+Sample output (as hexdump -C bytes)
+```
+00000000  02 04 6b 65 79 31 06 76  61 6c 75 65 31 04 6b 65  |..key1.value1.ke|
+00000010  79 32 06 76 61 6c 75 65  32                       |y2.value2|
+00000019
+```
+
+### New embedding API entrypoint
+
+We want at least 2 ways to pass data in: either ask the runtime to open the file, or give it a pointer to the data in memory. Also, we need some way to cleanup.
+
+```
+struct MonovmRuntimeConfigArguments {
+  uint32_t kind; // 0 = Path of runtimeconfig.bin file, 1 = pointer to the blob data, >= 2 reserved
+  union {
+    struct {
+      const char *path; // null terminated absolute path
+    } name;
+    struct {
+      const char *data;
+      uint32_t data_len;
+    } data;
+  } runtimeconfig;
+};
+
+typedef void (*MonovmRuntimeConfigArgumentsCleanup)(MonovmRuntimeConfigArguments *args, void* user_data);
+
+MONO_API void
+monovm_runtimeconfig_initialize (MonovmRuntimeConfigArguments *args, MonovmRuntimeConfigArgumentsCleanup cleanup_fn, void* user_data);
+```
+
+This declaration should live in the unstable header. `monovm_runtimeconfig_initialize` should be called after `monovm_initialize` but before starting the runtime.
```

Comment:
I thought we ended up going with the other way around - `monovm_runtimeconfig_initialize` before `monovm_initialize` before starting the runtime? Check to make sure this is consistent with the code, but as long as it is we should be fine.

Reply:
It is before. Let me fix this